### PR TITLE
perf(platform-browser): avoid intermediate arrays in server transition

### PR DESCRIPTION
**Pull request by crisbeto**
The server transition initializer looks for some `style` tags and clears them based on their `ng-transition` ID. The way we currently have the logic creates a couple of intermediate arrays just so we can call `forEach` at the end.

These changes use a regular `for` loop with an `if` statement instead. This isn't a _massive_ performance improvement, but the logic does run during app initialization which is performance-sensitive and it's an easy change to make on our end.

### CODE REVIEW

File: packages/platform-browser/src/browser/server-transition.ts [hunk 1]
```diff
@@ -21,10 +21,13 @@ export function appInitializerFactory(transitionId: string, document: any, injec
     // the server.
     injector.get(ApplicationInitStatus).donePromise.then(() => {
       const dom = getDOM();
-      const styles: any[] =
-          Array.prototype.slice.apply(document.querySelectorAll(`style[ng-transition]`));
-      styles.filter(el => el.getAttribute('ng-transition') === transitionId)
-          .forEach(el => dom.remove(el));
+      const styles: HTMLCollectionOf<HTMLStyleElement> =
+          document.querySelectorAll('style[ng-transition]');
+      for (let i = 0; i < styles.length; i++) {
```
Comment (crisbeto) [inline packages/platform-browser/src/browser/server-transition.ts:26]: I didn't use a `for...of` here, because I remember there being some sort of issue when it's used on a non-array iterable when downleveled to ES5.

File: packages/platform-browser/src/browser/server-transition.ts [hunk 2]
```diff
@@ -21,10 +21,13 @@ export function appInitializerFactory(transitionId: string, document: any, injec
     // the server.
     injector.get(ApplicationInitStatus).donePromise.then(() => {
       const dom = getDOM();
-      const styles: any[] =
-          Array.prototype.slice.apply(document.querySelectorAll(`style[ng-transition]`));
-      styles.filter(el => el.getAttribute('ng-transition') === transitionId)
-          .forEach(el => dom.remove(el));
+      const styles: HTMLCollectionOf<HTMLStyleElement> =
+          document.querySelectorAll('style[ng-transition]');
```
Comment (dario-piotrowicz) [inline packages/platform-browser/src/browser/server-transition.ts:25]: sorry if it's a silly question.... I was just wondering, couldn't we do a more specific `querySelectorAll` and avoid the `if` in the loop? 

like so:
```ts
const styles: HTMLCollectionOf<HTMLStyleElement> =
    document.querySelectorAll(`style[ng-transition="${transitionId}"]`);
```

and then:
```ts
for(let i = 0; i < styles.length; i++){
 dom.remove(styles[i]);
}
```

that would generate a smaller `styles` array and no `if`, wouldn't that be better?

File: packages/platform-browser/src/browser/server-transition.ts [hunk 3]
```diff
@@ -7,7 +7,7 @@
  */
 
 import {DOCUMENT, ɵgetDOM as getDOM} from '@angular/common';
-import {APP_INITIALIZER, ApplicationInitStatus, Inject, InjectionToken, Injector, StaticProvider} from '@angular/core';
```
Comment (dario-piotrowicz) [inline packages/platform-browser/src/browser/server-transition.ts:10]: was the `Inject` not used? shouldn't this be part of a separate commit? :sweat_smile:

File: packages/platform-browser/src/browser/server-transition.ts (hunk 2 above)
Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:25]: Ah good point, not sure how I missed that 😅

File: packages/platform-browser/src/browser/server-transition.ts (hunk 3 above)
Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:10]: Yeah, it wasn't being used. I don't think it needs a separate commit since I'm editing the same file anyway.

Comment (dario-piotrowicz) [reply packages/platform-browser/src/browser/server-transition.ts:10]: ah ok I was saying that because once I was editing the same file and I've actually been asked to move an unrelated change like this into it's own file :sweat_smile:

**Review by dario-piotrowicz — APPROVED** [review]

**Comment by dario-piotrowicz** [comment]
( :arrow_up:  pretty useless approval but I thought, why not :sweat_smile: :slightly_smiling_face:  )

File: packages/platform-browser/src/browser/server-transition.ts [hunk 4]
```diff
@@ -21,10 +21,11 @@ export function appInitializerFactory(transitionId: string, document: any, injec
...
     injector.get(ApplicationInitStatus).donePromise.then(() => {
       const dom = getDOM();
-      const styles: any[] =
-          Array.prototype.slice.apply(document.querySelectorAll(`style[ng-transition]`));
-      styles.filter(el => el.getAttribute('ng-transition') === transitionId)
-          .forEach(el => dom.remove(el));
+      const styles: HTMLCollectionOf<HTMLStyleElement> =
+          document.querySelectorAll(`style[ng-transition="${transitionId}"]`);
+      for (let i = 0; i < styles.length; i++) {
+        dom.remove(styles[i]);
```
Comment (alfaproject) [inline packages/platform-browser/src/browser/server-transition.ts:27]: Shouldn't this be?
```suggestion
        dom.remove(styles[0]);
```
(because the loop is moving forward instead of backwards)

Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:27]: The list here doesn't update so what I have above is correct (see https://codepen.io/crisbeto/pen/VwbNzNq?editors=1011). What you're describing is only necessary when reading `element.children` or `element.childNodes` as elements are being removed.

File: packages/platform-browser/src/browser/server-transition.ts [hunk 5]
```diff
@@ -21,10 +21,11 @@ export function appInitializerFactory(transitionId: string, document: any, injec
     // the server.
     injector.get(ApplicationInitStatus).donePromise.then(() => {
       const dom = getDOM();
-      const styles: any[] =
-          Array.prototype.slice.apply(document.querySelectorAll(`style[ng-transition]`));
-      styles.filter(el => el.getAttribute('ng-transition') === transitionId)
-          .forEach(el => dom.remove(el));
+      const styles: HTMLCollectionOf<HTMLStyleElement> =
+          document.querySelectorAll(`style[ng-transition="${transitionId}"]`);
```
Comment (AndrewKushnir) [inline packages/platform-browser/src/browser/server-transition.ts:25]: Great optimization (which also makes the code easier to read!), thanks @crisbeto 👍 

One thing that I wanted to mention is that we don't seem to have any sanitization/pre-processing of the `transitionId` value, i.e. we use whatever was provided as is. Previously it was not a big concern in this part of the code, since we just checked against that value, but now we use that as a part of a selector. I've tested the code a bit with different values and it seems to be working properly (including different quotes and escaping), but I wanted to mention that anyway in case you see any potential for regression here and/or if we should add some tests to verify potential corner cases.

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:391068640:BASE:391068657:1629132455506:827e817a).

File: packages/platform-browser/src/browser/server-transition.ts (hunk 5 above)
Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:25]: I don't think that we need to worry about it, because the transition ID is something we generate and I believe it's just a number.

Comment (AndrewKushnir) [reply packages/platform-browser/src/browser/server-transition.ts:25]: If I understand correctly, the transition id might be coming from the `BrowserModule.withServerTransition` calls ([see this example](https://github.com/angular/angular/blob/master/integration/platform-server/src/helloworld/app.ts#L17)), which developers can invoke (since it's a part of the public API).

Comment (alan-agius4) [reply packages/platform-browser/src/browser/server-transition.ts:25]: Indeed the transitionID is string which is controlled by the users.

**Comment by AndrewKushnir** [comment]
@crisbeto FYI presubmits are successful. I'm adding the "cleanup" label for now, feel free to remove it once   https://github.com/angular/angular/pull/43145#discussion_r689722622 is addressed.

**Comment by crisbeto** [comment]
I'll remove the cleanup label, because you mentioned that it works as expected even with unescaped quotes in https://github.com/angular/angular/pull/43145#discussion_r689697080.

File: packages/platform-browser/src/browser/server-transition.ts (hunk 5 above)
Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:25]: The comment above mentions that it works as expected even with edge-casey values so it should be fine.

Comment (AndrewKushnir) [reply packages/platform-browser/src/browser/server-transition.ts:25]: @crisbeto I did some manual testing, so I'm wondering if it'd be beneficial to add some tests?

Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:25]: As far as I can tell, this logic isn't tested at all so we'd need to figure out how to test it first. Either way, I didn't intend for this to turn into such a discussion. The main performance optimization is avoiding the extra arrays. If you think that the selector is such a problem, I can switch it back to the old approach that was using `getAttribute` inside the loop.

Comment (AndrewKushnir) [reply packages/platform-browser/src/browser/server-transition.ts:25]: I don't want to block/delay this optimization, so if there is no (or very very low) risk of a regression caused by selector modification, let's proceed with the change.

Comment (crisbeto) [reply packages/platform-browser/src/browser/server-transition.ts:25]: I think the only risk for a regression is if the transition IDs have double quotes in them, but based on your testing it sounds like that is handled by the browser as well.
